### PR TITLE
WIP Add report scenarios

### PR DESCRIPTION
**Pull request by dominikhajduk**

### CODE REVIEW

File: taskmanager/features/report.feature [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  I want to generate various of reports
+
+  Scenario: Generate report for TODO tasks
+    Given There is no report
+    And there are "2" tasks with "TODO" status
+    When I generate report for "TODO" tasks
+    Then there should be "1" report
+
+
+#  Scenario: Generate report for open tasks
```
Comment (partikus) [inline taskmanager/features/report.feature]: please use `tags` instead of commenting scenarios

File: taskmanager/features/report.feature [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# language: en
+Feature: Generate a report
+  In order to export tasks with details
+  As a user or system
+  I want to generate various of reports
+
+  Scenario: Generate report for TODO tasks
+    Given There is no report
```
Comment (partikus) [inline taskmanager/features/report.feature]: `Given There is no reports` - plural v.

File: taskmanager/features/bootstrap/FeatureContext.php [hunk 3]
```diff
@@ -15,6 +15,9 @@ class FeatureContext implements Context
     /** @var \App\Application\Task\TaskRegistry */
     private $taskRegistry;
 
+    /** @var \App\Application\Report\ReportRegistry */
```
Comment (partikus) [inline taskmanager/features/bootstrap/FeatureContext.php]: we should not put everything under a single context. 
http://behat.org/en/latest/user_guide/context.html 
maybe sth like that `ReportContext` and `TaskContext` ?

File: taskmanager/features/report.feature [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    And there are "2" tasks with "TODO" status
+    When I generate report for "TODO" tasks
```
Comment (partikus) [inline taskmanager/features/report.feature]: How does it work? Any more detailed reqs for reports?
IMHO reports should present data like: 
- unfinished
- done 
- inprogress
- some charts and comparision between them

what was your a first idea when you said "Reports feature"? Could you explain to us that bit more?

File: taskmanager/src/Application/Report/Report.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        return $this->id;
+    }
+
+    public function getName(): string
+    {
+        return $this->name;
+    }
+
+    public function addElement(Task $task)
```
Comment (partikus) [inline taskmanager/src/Application/Report/Report.php]: `addTask` ?

File: taskmanager/src/Application/Report/Report.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        $this->elements[] = $task;
+    }
+
+    public function getElements(): array
```
Comment (partikus) [inline taskmanager/src/Application/Report/Report.php]: `getTasks`

**Review by partikus — CHANGES_REQUESTED** [review]
@dominikhajduk would be better if you provide more scenarios' examples to clarify what should be done and whether an implementation is appropriate to reqs. 

The PoC is fine. My suggestions:
1. Improve BDD scenarios
2. Think about the solution that will able to handle all of your requirements. 
3. Provide scenario-by-scenario new kind of reports.

File: taskmanager/features/bootstrap/ReportContext.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    /**
+     * @When I generate report with criteria :arg1
+     */
+    public function iGenerateReportWithCriteria($criteria)
+    {
+        /** @var Criteria $criteria */
+        $criteria = Criteria::{$criteria}();
+        $report = new \App\Application\Report\Report($criteria->__toString());
+
+        switch ($criteria->getCriteria()) {
```
Comment (dominikhajduk) [inline taskmanager/features/bootstrap/ReportContext.php]: probably need switch to some strategy inside. So create report where first param is `Criteria` based on that create report.

File: taskmanager/features/report.feature [hunk 8]
```diff
@@ -0,0 +1,29 @@
+# language: en
+Feature: Generate a report
+  In order to export tasks with details
+  As a user or system
+  I want to generate various of reports
+
+  Scenario: Generate report for TODO tasks
+    Given There is no reports
+    And there are "2" tasks with "TODO" status
+    When I generate report with criteria "unfinished"
```
Comment (partikus) [inline taskmanager/features/report.feature]: Maybe would better to define criteria as a list.
```gherkin
When I generate report with the following criteria:
  | Key        | Value          |
  | status     | unfinished  |
```
etc. thanks to that you would be able  to define more than one condition

File: taskmanager/src/Application/Report/Criteria.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
+<?php declare(strict_types=1);
+
+namespace App\Application\Report;
+
+class Criteria
```
Comment (partikus) [inline taskmanager/src/Application/Report/Criteria.php:5]: I'm not sure whether criteria should be VO in such form. `Criteria` object accepts only status of the task. IMHO it's not enough even at the beginning. How to handle a case e.g. `createdAt BETWEEN(2017-01-01, 2017-02-02)` ?

**Review by partikus — CHANGES_REQUESTED** [review]

File: taskmanager/src/Application/Report/Criteria.php (hunk 9 above)
Comment (dominikhajduk) [reply taskmanager/src/Application/Report/Criteria.php:5]: @partikus true there must me option to see not only all todo but also this week todo.

**Comment by dominikhajduk** [comment]
Added scenarios which ar not yet covered with the code https://github.com/PGS-php/upskilling-2nd-edition/pull/16/commits/15710511474f6af76ce86e5a5638a679c76e7878

File: taskmanager/features/bootstrap/ReportContext.php [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    /**
+     * @Given the following tasks exist:
+     */
+    public function theFollowingTasksExist(TableNode $table)
+    {
+        $this->taskRegistry = new TaskRegistry();
+        foreach ($table as $item) {
+            //@TODO: implement other statuses
```
Comment (dominikhajduk) [inline taskmanager/features/bootstrap/ReportContext.php:64]: This probably waits for @mateuszmaslanka PR 🤔

File: taskmanager/features/bootstrap/ReportContext.php [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+            $status = Status::toDo();
+            //@TODO: implement create and update date
```
Comment (dominikhajduk) [inline taskmanager/features/bootstrap/ReportContext.php]: Implemented here https://github.com/PGS-php/upskilling-2nd-edition/pull/19 🥇

File: taskmanager/spec/Application/Report/CriteriaSpec.php [hunk 12]
```diff
@@ -0,0 +1,62 @@
+<?php declare(strict_types=1);
+
+namespace spec\App\Application\Report;
+
+use App\Application\Report\Criteria;
+use PhpSpec\ObjectBehavior;
+
+class CriteriaSpec extends ObjectBehavior
```
Comment (dominikhajduk) [inline taskmanager/spec/Application/Report/CriteriaSpec.php:8]: Need to be updated!

File: taskmanager/spec/Application/Report/ReportSpec.php [hunk 13]
```diff
@@ -0,0 +1,37 @@
+<?php declare(strict_types=1);
+
+namespace spec\App\Application\Report;
+
+use App\Application\Report\Report;
+use PhpSpec\ObjectBehavior;
+use Ramsey\Uuid\UuidInterface;
+
+class ReportSpec extends ObjectBehavior
```
Comment (dominikhajduk) [inline taskmanager/spec/Application/Report/ReportSpec.php:10]: Need to be updated!

File: taskmanager/src/Application/Report/Report.php [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+    public function getName(): string
+    {
+        return $this->name;
+    }
+
+    public function generate(TaskRegistry $tasks)
+    {
+        //@TODO: foreach criteria get tasks and add to elements or tasks
+        // how to decide if assigne tasks or elements like Efficiency | 33,3% hmm.. ?
```
Comment (dominikhajduk) [inline taskmanager/src/Application/Report/Report.php]: @tarnawski @partikus @mateuszmaslanka @dgrochowski Hmm.. that's the most important question for that PR. And no idea how to solve 🤔 Comments more than appreciated.

File: taskmanager/features/bootstrap/ReportContext.php [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+    private $report;
+
+    /**
+     * Initializes context.
+     *
+     * Every scenario gets its own context instance.
+     * You can also pass arbitrary arguments to the
+     * context constructor through behat.yml.
+     */
+    public function __construct()
```
Comment (tarnawski) [inline taskmanager/features/bootstrap/ReportContext.php]: We need this empty constructor?

File: taskmanager/src/Application/Report/Report.php (hunk 14 above)
Comment (tarnawski) [reply taskmanager/src/Application/Report/Report.php]: Maybe we should simplify the logic of generating reports and create one type of report contains task filtered by criteria.

File: taskmanager/src/Infrastructure/InMemory/ReportRegistry.php [hunk 16]
```diff
@@ -0,0 +1,32 @@
+<?php declare(strict_types=1);
```
Comment (tarnawski) [inline taskmanager/src/Infrastructure/InMemory/ReportRegistry.php:1]: I'm not sure that we need store reports - generate `in fly` should be enough

File: taskmanager/src/Application/Report/Criteria.php (hunk 9 above)
Comment (tarnawski) [reply taskmanager/src/Application/Report/Criteria.php:5]: Add function can get name and object of single criteria element like bellow:
```php
public function add(string $name, CriteriaElementInterface $criteriaElement);
```

**Review by tarnawski — CHANGES_REQUESTED** [review]

File: taskmanager/features/bootstrap/ReportContext.php (hunk 15 above)
Comment (dominikhajduk) [reply taskmanager/features/bootstrap/ReportContext.php]: @tarnawski no created automatically.

File: taskmanager/src/Infrastructure/InMemory/ReportRegistry.php (hunk 16 above)
Comment (dominikhajduk) [reply taskmanager/src/Infrastructure/InMemory/ReportRegistry.php:1]: @tarnawski removed.

**Comment by dominikhajduk** [comment]
Simplified report generation to just list tasks. Stil few paces need to be finished but concept should be clear for review.

**Comment by dominikhajduk** [comment]
abandoned project

**Pull request closed without merging**